litestar/utils/predicates.py

```diff
@@ -212,6 +213,28 @@ def is_non_string_sequence(annotation: Any) -> TypeGuard[Sequence[Any]]:
         return False
 
 
+def is_non_string_non_mapping_iterable(annotation: Any) -> TypeGuard[Iterable[Any]]:
```

Comment:
Why does this specifically exclude `Mapping`? Wouldn't it make more sense to adapt the existing `is_non_string_sequence` and create a `is_non_string_iterable`? `Mapping` is an iterable after all, and if we're going to be broad in what we accept here, why specifically exclude it?

Reply:
from collections.abc import Mapping, Iterable
issubclass(Mapping, Iterable)
True

Reply:
It's been two weeks, I'll need to take a look.

Reply:
It's because `Object` (python type `Mapping`) is encoded differently. The idea was to support python types that are equivalent of `Array`. If we widen accepted types from `Sequence` to `Iterable` to support `Array` we must exclude `Mapping`.

Also HTTP allows Arrays and Objects with only Scalar values, and python `Mapping` element is complex type `tuple`.

Reply:
> It's been two weeks, I'll need to take a look.

Don't have much time outside work at the moment, sorry :)



> It's because `Object` (python type `Mapping`) is encoded differently. The idea was to support python types that are equivalent of `Array`. If we widen accepted types from `Sequence` to `Iterable` to support `Array` we must exclude `Mapping`.
> 
> Also HTTP allows Arrays and Objects with only Scalar values, and python `Mapping` element is complex type `tuple`.

Not sure I follow your reasoning here. If we allow `Mappings` where we allow `Iterable`, I wouldn't expect any issues, since you *can* iterate over a mapping, hence making it an iterable. As you pointed out in [this](https://github.com/litestar-org/litestar/issues/3631#issuecomment-2231645694) discussion, things aren't as strict as they should be anyway, so the line to draw is kinda arbitrary. Explicitly excluding `Mapping` would just move the arbitrary line somewhere else, and IMO it's more intuitive to just treat all non-string iterables the same here.

In any case, why can't we reuse `is_non_string_iterable` for this?

Reply:
> Don't have much time outside work at the moment, sorry :)

No problem, that's life 🤷

> I wouldn't expect any issues, since you can iterate over a mapping, hence making it an iterable

It's not about Markov principle, it's about how these types are mapped to OpenAPI and which parses are used on incoming requests.
It's clear to me that `Iterable[Scalar]` should be represented in OpenAPI as `array` and should make `ParameterDefinition.is_sequence` be `True`.
But `Mapping` is really `Iterable[tuple]` - what would it map to? Specifications don't support arrays of non-scalar types.

> As you pointed out in https://github.com/litestar-org/litestar/issues/3631#issuecomment-2231645694 discussion, things aren't as strict as they should be anyway, so the line to draw is kinda arbitrary

I said things aren't as intuitive as they should be :)

If you mean https://github.com/litestar-org/litestar/pull/3632#issuecomment-2230898096, that's completely different story. My thinking was going even further than this PR.

I thought of introducing a concept of `ArrayLike`, which would encompass python types that correspond to HTTP Array.
`is_non_string_non_mapping_iterable` could be renamed to `is_array_like`, and would replace `is_non_string_sequence` in most (perhaps all) places, since this is what its implementation suggest it's trying to do.

Reply:
> It's not about Markov principle

OK, it is all about Markov principle :) only we're handling a dual type system - python and HTTP/OpenAPI.